File: DesignPatterns/Behavioral/TemplateMethod/StoreClient.java
abstract class OrderProcessTemplate {
  public boolean isGift;

  public abstract void doSelect();

  public abstract void doPayment();

  public final void giftWrap() {
    try {
      System.out.println("Giftwrap successfull");
    } catch(Exception e) {
      System.out.println("Giftwrap unsuccessfull");
    }
  }

  public abstract void doDelivery();

  public final void processOrder(boolean isGift) {
    doSelect();
    doPayment();
    if(isGift){
      giftWrap();
    }
    doDelivery();
  }
}


class NetOrder extends OrderProcessTemplate {

  public void doSelect() {
    System.out.println("Item added to online shopping cart");
    System.out.println("Get giftWrap preference");
    System.out.println("Get delivery address");
  }

  public void doPayment() {
    System.out.println("Online payment through Netbanking, card or paytm");
  }

  public void doDelivery() {
    System.out.println("Ship the item through post to delivery address");
  }
}

class StoreOrder extends OrderProcessTemplate {

  public void doSelect() {
    System.out.println("Customer chooses the item from the shelf");
  }

  public void doPayment() {
    System.out.println("Pays at counter through cash/POS");
  }

  public void doDelivery() {
    System.out.println("Item deliverd to in delivery counter");
  }
}


public class StoreClient {
  public static void main(String[] args) {
    OrderProcessTemplate netOrder = new NetOrder();
    netOrder.processOrder(true);
    System.out.println("***********");
    OrderProcessTemplate storeOrder = new StoreOrder();
    storeOrder.processOrder(true);
  }
}

File: DesignPatterns/Behavioral/TemplateMethod/Ruby/store_client.rb
class OrderProcessTemplate
  def isGift
    raise NotImplementedError
  end

  def do_select
    raise NotImplementedError
  end

  def do_payment
    raise NotImplementedError
  end

  def gift_wrap
    puts "Giftwrap successfull"
  rescue
    puts "Giftwrap unsuccessfull"
  end

  def do_delivery
    raise NotImplementedError
  end

  def process_order(gift)
    do_select
    do_payment
    gift_wrap if gift
    do_delivery
  end
end

class NetOrder < OrderProcessTemplate
  def do_select
    puts "Item added to online shopping cart"
    puts "Get giftWrap preference"
    puts "Get delivery address"
  end

  def do_payment
    puts "Online payment through Netbanking, card or paytm"
  end

  def do_delivery
    puts "Ship the item through post to delivery address"
  end
end

class StoreOrder < OrderProcessTemplate
  def do_select
    puts 'Customer chooses the item from the shelf'
  end

  def do_payment
    puts 'Pays at counter through cash/POS'
  end

  def do_delivery
    puts 'Item deliverd to in delivery counter'
  end
end

net_order = NetOrder.new
net_order.process_order(true)
puts "********"
store_order = StoreOrder.new
store_order.process_order(false)